### ISSUE TITLE
FAPI Crashes in test_policy_action

### ISSUE THREAD

**Issue by williamcroberts** [tpm2-software/tpm2-pytss#169]
I am on tpm2-tss version 3.1.0, Ubuntu 20.04, Python 3.8.5 and CFFI 1.14.5. I was initially thinking that this crash had something to do with Callback fn signature mismatches, but the testing TSS version is 3.0.3 which has those changes, so I am not sure.

```
================================================== test session starts ===================================================
platform linux -- Python 3.8.5, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /home/wcrobert/workspace/tpm2-pytss
plugins: cov-2.12.1, forked-1.3.0, xdist-2.2.1
gw0 [53] / gw1 [53] / gw2 [53] / gw3 [53]
................................................Fatal Python error: Aborted

Thread 0x00007f02a8ff3700 (most recent call first):
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 400 in read
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 432 in from_io
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 967 in _thread_receiver
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 220 in run
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 285 in _perform_spawn

Current thread 0x00007f02aa0b7740 (most recent call first):
  File "/home/wcrobert/workspace/tpm2-pytss/tpm2_pytss/FAPI.py", line 331 in sign
  File "/home/wcrobert/workspace/tpm2-pytss/test/test_fapi.py", line 900 in test_policy_action
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/python.py", line 183 in pytest_pyfunc_call
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/python.py", line 1641 in runtest
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/runner.py", line 162 in pytest_runtest_call
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/runner.py", line 255 in <lambda>
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/runner.py", line 311 in from_call
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/runner.py", line 254 in call_runtest_hook
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/runner.py", line 215 in call_and_report
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/runner.py", line 126 in runtestprotocol
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/runner.py", line 109 in pytest_runtest_protocol
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/wcrobert/.local/lib/python3.8/site-packages/xdist/remote.py", line 89 in run_one_test
  File "/home/wcrobert/.local/lib/python3.8/site-packages/xdist/remote.py", line 75 in pytest_runtestloop
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/main.py", line 323 in _main
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/main.py", line 269 in wrap_session
  File "/home/wcrobert/.local/lib/python3.8/site-packages/_pytest/main.py", line 316 in pytest_cmdline_main
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/wcrobert/.local/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/wcrobert/.local/lib/python3.8/site-packages/xdist/remote.py", line 244 in <module>
  File "/home/wcrobert/.local/lib/python3.8/site-packa.ges/execnet/gateway_base.py", line 1084 in executetask
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 220 in run
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 285 in _perform_spawn
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 267 in integrate_as_primary_thread
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 1060 in serve
  File "/home/wcrobert/.local/lib/python3.8/site-packages/execnet/gateway_base.py", line 1554 in serve
  File "<string>", line 8 in <module>
  File "<string>", line 1 in <module>
...[gw2] node down: Not properly terminated
F
replacing crashed worker gw2
gw0 [53] / gw1 [53] / gw4 ok / gw3 [53]  
======================================================== FAILURES ========================================================
___________________________________________________ test/test_fapi.py ____________________________________________________
[gw2] linux -- Python 3.8.5 /usr/bin/python3
worker 'gw2' crashed while running 'test/test_fapi.py::TestFapi::test_policy_action'
================================================ short test summary info =================================================
FAILED test/test_fapi.py::TestFapi::test_policy_action
============================================== 1 failed, 52 passed in 3.72s ==============================================
```

**Comment by williamcroberts** [comment]
@joholl

**Comment by joholl** [comment]
TL;DR you need a bleeding edge TSS

This is most certainly a known problem. The issue is that using PolicyAction and Fapi_Sign() result in a double free, see tpm2-software/tpm2-tss/issues/2089. This is fixed upstream, so be sure to use a recent master TSS build. Note that the `test_policy_signed` is skipped for TSS versions below `3.1.0`.

**Comment by williamcroberts** [comment]
@joholl, seems like the checks are wrong, shouldn't we have the check `<= 3.1.0` (less than equal to, not less than). It's executing for 3.1.0 which seems to have the bugs.

**Comment by williamcroberts** [comment]
Then the other issue, a user installs this with a bad fapi version, should we check in the broken functions and raise a runtime error so the user isn't reporting these? Provide them the details of what was broken and corrective actions?

**Comment by joholl** [comment]
You're totally right! My bad.

And yes, we should throw a RuntimeException for the double frees (and maybe log a warning when Fapi_Quote() is called), good call!

**Comment by williamcroberts** [comment]
@joholl did we get anywhere with these? I don't recall seeing a PR offhand.

**Comment by williamcroberts** [comment]
So I updated the test portion to be if version is less than or equal to 3.1.0 in : https://github.com/tpm2-software/tpm2-pytss/commit/ffcb39747a61d38af483349dbaee32b26ffb356b

**Comment by williamcroberts** [comment]
So TODO is that actual API Runtime Error

**Comment by joholl** [comment]
That is an actual RuntimeError that happens if using a PolicyAction key with Fapi_Sign. I didn't have the time yet to add warnings to the python FAPI (and double check all conditional skips in the fapi tests). I will tackle that soon and keep you notified!
